packages/aws_common/lib/src/http/aws_http_client_io.dart

```diff
@@ -0,0 +1,500 @@
+/// - On the Web, this uses the browser's `fetch` API.
+/// {@endtemplate}
+class AWSHttpClientImpl extends AWSHttpClient {
+  /// {@macro aws_common.http.http_client_impl}
+  AWSHttpClientImpl() : super.protected();
+
+  /// The underlying `dart:io` HTTP/1.1 client.
+  HttpClient? _inner;
+
+  final _http2Connections = <String, ClientTransportConnection>{};
+
+  static final AWSLogger _logger = AWSLogger().createChild('AWSHttpClient');
+
+  void _setBadCertificateCallback(
+    HttpClient client,
+    BadCertificateCallback callback,
+  ) {
+    client.badCertificateCallback = (cert, host, port) {
+      final certificate = X509Certificate(
+        der: cert.der,
+        pem: cert.pem,
+        sha1: cert.sha1,
+        subject: cert.subject,
+        issuer: cert.issuer,
+        startValidity: cert.startValidity,
+        endValidity: cert.endValidity,
+      );
+      return onBadCertificate(certificate, host, port);
+    };
+  }
+
+  @override
+  set onBadCertificate(BadCertificateCallback callback) {
+    if (_inner != null) {
+      _setBadCertificateCallback(_inner!, callback);
+    }
+    super.onBadCertificate = callback;
+  }
+
+  /// Sends an HTTP/1.1 request using `dart:io`.
+  Future<void> _sendH1({
+    required AWSLogger logger,
+    required AWSBaseHttpRequest request,
+    required CancelableCompleter<AWSBaseHttpResponse> completer,
+    required Future<void> cancelTrigger,
+    required StreamSink<int> requestProgress,
+    required StreamSink<int> responseProgress,
+  }) async {
+    void Function()? onCancel;
+    unawaited(
+      cancelTrigger.then((_) {
+        logger.verbose('Canceling request');
+        onCancel?.call();
+      }),
+    );
+    _inner ??= HttpClient();
+    _setBadCertificateCallback(_inner!, onBadCertificate);
+    if (completer.isCanceled) return;
+    final ioRequest = (await _inner!.openUrl(request.method.value, request.uri))
+      ..followRedirects = request.followRedirects
+      ..maxRedirects = request.maxRedirects;
+    if (request.hasContentLength) {
+      ioRequest.contentLength = request.contentLength as int;
+    }
+
+    if (completer.isCanceled) return;
+    onCancel = ioRequest.abort;
+
+    var requestBytesRead = 0;
+    request.headers.forEach(ioRequest.headers.set);
+    final response = await request.body
+        .tap(onDone: requestProgress.close, (chunk) {
+          requestBytesRead += chunk.length;
+          requestProgress.add(requestBytesRead);
+        })
+        .takeUntil(cancelTrigger)
+        .pipe(ioRequest) as HttpClientResponse;
+
+    if (completer.isCanceled) return;
+    final bodyController = StreamController<List<int>>(sync: true);
+    onCancel = () {
+      logger.verbose('Detaching socket');
+      if (!bodyController.isClosed) {
+        bodyController
+          ..addError(const CancellationException())
+          ..close();
+      }
+      response.detachSocket().then((socket) {
```

Comment:
nice, great job re-registering onCancel cb to call appropriate io method depending on state of response. There was some discuss of this in package:http and this looks correct to me.

Reply:
Yeah I had bumped into that too - probably wouldn't have figured that out otherwise.